### DISCUSSION TITLE
Proposal: Allow try to modify using/switch statements

### DISCUSSION THREAD

**Discussion by heaton84** [dotnet/csharplang#2655] [Language Ideas]
This is a simple reduction in braces request. Often times, I find that spinning up an unmanaged resource requires a coinciding try/catch handler. For example:

```csharp
try
{
    using (DatabaseInterface db = new DatabaseInterface())
    {
        db.Connect();
        db.DoSomeStuff();
    }
}
catch (Exception ex)
{
    Log("Database failed to DoSomeStuff: {0}{1}", ex.Exception, ex.StackTrace);
}
```
Can be shortened a few lines as follows:

```csharp
try using (DatabaseInterface db = new DatabaseInterface())
{
        db.Connect();
        db.DoSomeStuff();
}
catch (Exception ex)
{
    Log("Database failed to DoSomeStuff: {0}{1}", ex.Exception, ex.StackTrace);
}
```
This can also be extended to a `switch` statement:
```csharp
try switch (workType)
{
    case 0:
        workerClass.Zig();
        break;

    case 1:
        workerClass.Zag();
        break;
        
    default:
        throw new NotImplementedException(someVariable.ToString());
}
catch
{

}
```

I feel like this helps avoid needless syntax, while still clearly describing what is happening.

I don't see any additional complexities being introduced. Offhand, I can't think of any other keywords that require braces that this could be extended to beyond using and switch.

**Comment by TheUnlocked** [comment]
A more general solution would be to just allow `try`/`catch` to be used on a single line and not require braces, like with `if`/`else`:
```csharp
if (pred)
    doSomething();
else
    doSomethingElse();
```

**Comment by yaakov-h** [comment]
@TheUnlocked allowing that is basically impossible due to nesting and multiple catches, and has been discussed extensively already in other issues.

**Comment by YairHalberstadt** [comment]
@yaakov-h

Why is that any different to

```csharp
if (a)
if (b)
   M1();
else
    M2();
```
Where its ambiguous which if the else is attached to.

**Comment by yaakov-h** [comment]
@YairHalberstadt That reads ambiguously, but is actually well-defined. The `else` gets attached to the inner-most `if`.

On the other hand, how would you well-define this case?:

```csharp
try
try
  M();
catch (IOException e) { }
catch (SystemException e) { }
finally { }
```

**Comment by YairHalberstadt** [comment]
@yaakov-h
Same. Arbitrarily define that everything gets attached to the innermost try.

**Comment by HaloFour** [comment]
@YairHalberstadt 

It can't, every `try` needs a corresponding `catch` or `finally`, which means that you'd be assigning some number of those clauses to the nested `try` statements that are not innermost.  Even if the behavior of the compiler could be well defined it would not be intuitive, it's simply not worth it.

Bjarne Stroustroup explicitly called this out in 16.8 Appendix D of [Exception Handling for C++](http://www.stroustrup.com/except89.pdf):

> Allowing exception handlers to be attached to blocks only and not to simple statements simplifies syntax analysis (both for humans and computers) where several exceptions are caught and where nested exception handlers are considered (see Appendix E). For example, assuming that we allowed handlers to be attached to any statement we could write:
> 
> ```
> try try f(); catch (x) { ... } catch (y) { ... } catch (z) { ... }
> ```
> 
> The could be interpreted be in at least three ways:
> 
> ```
> try { try f(); catch (x) { ... } } catch (y) { ... } catch (z) { ... }
> try { try f(); catch (x) { ... } catch (y) { ... } } catch (z) { ... }
> try { try f(); catch (x) { ... } catch (y) { ... } catch (z) { ... } }
> ```
> 
> There seems to be no reason to allow these ambiguities even if there is a trivial and systematic way for a parser to chose one interpretation over another. Consequently, a `{` is required after a `try` and a matching `}` before the first of the associated sequence of `catch` clauses.

**Comment by YairHalberstadt** [comment]
@HaloFour
I would simply make that an error.

**Comment by HaloFour** [comment]
@YairHalberstadt 

> I would simply make that an error.

Make what an error?  The nesting of `try` statements without blocks?

**Comment by YairHalberstadt** [comment]
@HaloFour
Yes

**Comment by YairHalberstadt** [comment]
Though that would make it tricky to in the future add catches to using statements, since it would cause the following code to become an error:

```csharp
try 
    using(...) 
        M();
catch {}
```

**Comment by HaloFour** [comment]
@YairHalberstadt 

Maybe, I'd be curious to what you'd have to do to the spec to make having `try try` illegal but `try anything else` legal.

And frankly the savings here is minimal.  You save two characters and an optional level of nesting.  The IDE writes most of that for you.

You can already do `using (...) try { ... } catch { ... }`.

**Comment by yaakov-h** [comment]
```csharp
try
    using (foo)
        try
            catch (Exception ex) { }
catch (Exception ex) // oops!
```

**Comment by YairHalberstadt** [comment]
@yaakov-h

That's a good point. I retract.

**Comment by heaton84** [comment]
I feel like the discussion got a little off the rails here. Trying to get back on point:

@HaloFour 

Your `using (...) try { ... } catch { ... }` example has one flaw... the instantiation of the object in the using block can throw an exception, requiring an outer try block to handle any errors in that case.


@yaakov-h 

Your example only breaks if we remove brackets from try/catch. Again, that is not my goal here.


While I will concede that the savings are minimal, I argue that there is no downside to the approach aside from (minimal) additional complications to the language. Effectively, we could get this monstrosity:

```csharp
try using (Foo bar = new Foo()) switch (bar.getFooBar())
{
    case Foo.Rod:
        break;

    case Foo.Stick:
        break;
}
catch (Exception ex)
{
}
```

The heart of the issue is being able to re-use existing syntax of statements like `try` and `switch`.

**Comment by DaZombieKiller** [comment]
If the issue is the extra level of nesting, then why not use a `using` declaration?

```CSharp
try
{
    using var db = new DatabaseInterface();
    db.Connect();
    db.DoSomeStuff();
}
catch (Exception ex)
{
    Log("Database failed to DoSomeStuff: {0}{1}", ex.Exception, ex.StackTrace);
}
```

**Comment by heaton84** [comment]
@DaZombieKiller 
I was not aware that was in the pipeline. Thanks for the lesson. When was that introduced? I can't find anything in the documentation.

**Comment by DaZombieKiller** [comment]
@heaton84 It's a C# 8 feature: https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#using-declarations
